[GSB] Only use matching superclass constraints in enumerated requirements

DougGregor opened the pull request:
Fixes SR-5165 / rdar://problem/32658705.

DougGregor committed:
[GSB] Only use matching superclass constraints in enumerated requirements.

Fixes SR-5165 / rdar://problem/32658705.

DougGregor commented:
@swift-ci please smoke test and merge

slavapestov commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -5092,14 +5095,21 @@ void GenericSignatureBuilder::enumerateRequirements(llvm::function_ref<
+        getBestConstraintSource<Type>(equivClass->superclassConstraints,
+                        [&](const Type &type) {
+                          return type->isEqual(equivClass->superclass) ||
+                            equivClass->superclass->hasError();
```
What's the significance of hasError() here?

DougGregor replied:
We may not have an exact match if there are errors. I've decided that I don't like this approach and will improve on this a bit.

DougGregor commented:
Ugh, bad test case merge

DougGregor committed:
[GSB] Fall back to "abstract" requirements when we don't find a match.

The previous commit tightened up our choice of requirement sources for
superclass and layout constraints, but always assumed a match
existed. In cases where it does not, don't fail; rather, use an
abstract source, which we do elsewhere when we're synthesizing
requirements.

This is a defensive approach to cope with cases where we might not
have precisely the same value within one of our recorded
constraints. This currently only occurs when there are errors (hence
the elimination of the odd "hasError()"), but will come up more often
when we're deriving information from a combination of constraints.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge
